crates/isograph_compiler/src/watch.rs

```diff
@@ -52,20 +64,106 @@ pub async fn handle_watch_command(
+) -> Option<Vec<SourceFileEvent>> {
+    let mut source_file_events = vec![];
+    for event in events {
+        if let Some(source_file_event) = match event.kind {
+            EventKind::Create(create_kind) => {
+                process_create_event(config, create_kind, &event.paths)
+            }
+            EventKind::Modify(modify_kind) => {
+                process_modify_event(config, modify_kind, &event.paths)
             }
+            EventKind::Remove(remove_kind) => {
+                process_remove_event(config, remove_kind, &event.paths)
+            }
+            _ => None,
+        } {
+            source_file_events.push(source_file_event);
         }
     }
-    false
+    if source_file_events.is_empty() {
+        None
+    } else {
+        Some(source_file_events)
+    }
+}
+
+fn process_create_event(
+    config: &CompilerConfig,
+    create_kind: CreateKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match create_kind {
+        CreateKind::File if paths.len() == 1 => changed_file_kind(config, &paths[0])
+            .map(|file_kind| (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)),
+        _ => None,
+    }
+}
+
+fn process_modify_event(
+    config: &CompilerConfig,
+    modify_kind: ModifyKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match modify_kind {
+        ModifyKind::Data(_) if paths.len() == 1 => changed_file_kind(config, &paths[0])
+            .map(|file_kind| (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)),
+        ModifyKind::Name(rename_mode) => {
+            match rename_mode {
+                // This event could be fired once on delete or twice on rename
+                RenameMode::Any if paths.len() == 1 => {
+                    changed_file_kind(config, &paths[0]).map(|file_kind| {
+                        if paths[0].exists() {
+                            (SourceEventKind::CreateOrModify(paths[0].clone()), file_kind)
+                        } else {
+                            (SourceEventKind::Remove(paths[0].clone()), file_kind)
+                        }
+                    })
+                }
+                RenameMode::Both if paths.len() == 2 => {
+                    changed_file_kind(config, &paths[1]).map(|file_kind| {
+                        (
+                            SourceEventKind::Rename((paths[0].clone(), paths[1].clone())),
+                            file_kind,
+                        )
+                    })
+                }
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+}
+
+fn process_remove_event(
+    config: &CompilerConfig,
+    remove_kind: RemoveKind,
+    paths: &[PathBuf],
+) -> Option<SourceFileEvent> {
+    match remove_kind {
+        RemoveKind::File | RemoveKind::Folder if paths.len() == 1 => {
+            changed_file_kind(config, &paths[0])
+                .map(|file_kind| (SourceEventKind::Remove(paths[0].clone()), file_kind))
```

Comment:
Right now, if you remove a folder containing `iso` literals, it will remove that folder from `source_files`, but not remove any source files within that folder.

Perhaps we should have a `SourceEventKind::RemoveFile` and `SourceEventKind::RemoveFolder` variants, then if we process a `SourceEventKind::RemoveFolder` in `SourceFiles::handle_remove`, we iterate over the source files and remove the ones that start with that folder?

Likewise, we don't handle moves of folders correctly for the same reason